LoopKit/UserAlert.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  UserAlert.swift
+//  LoopKit
+//
+//  Created by Rick Pasetto on 4/8/20.
+//  Copyright © 2020 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+/// Protocol that describes any class that handles Alert posting.
+public protocol UserAlertHandler: class {
+    /// Schedule the given alert for posting.
+    func scheduleAlert(_ alert: UserAlert)
+    /// Unschedule any pending alerts with the given identifier.
+    func unscheduleAlert(identifier: UserAlert.Identifier)
+    /// Remove any alerts currently posted with the given identifier.  It ignores any pending alerts.
+    func cancelAlert(identifier: UserAlert.Identifier)
+}
+
+/// Protocol that describes something that can deal with a user's response to an alert.
+public protocol UserAlertResponder {
+    /// Acknowledge alerts with a given type identifier
+    func acknowledgeAlert(typeIdentifier: UserAlert.TypeIdentifier) -> Void
+}
+
+/// Structure that represents an Alert that needs to be shown to the User.
+public struct UserAlert {
+    /// Representation of an alert Trigger
+    public enum Trigger {
+        /// Trigger the alert immediately
+        case immediate
+        /// Delay triggering the alert by `interval`, but issue it only once.
+        case delayed(interval: TimeInterval)
+        /// Delay triggering the alert by `repeatInterval`, and repeat at that interval until cancelled or unscheduled.
+        case repeating(repeatInterval: TimeInterval)
+    }
+    /// Content of the alert, either for foreground or background alerts
+    public struct Content {
+        public let title: String
+        public let body: String
+        /// Should this alert be deemed "critical" for the User?  Handlers will determine how that is manifested.
+        public let isCritical: Bool
+        // TODO: when we have more complicated actions.  For now, all we have is "acknowledge".
+//        let actions: [UserAlertAction]
+        public let acknowledgeActionButtonLabel: String
+        public init(title: String, body: String, acknowledgeActionButtonLabel: String, isCritical: Bool = false) {
+            self.title = title
+            self.body = body
+            self.acknowledgeActionButtonLabel = acknowledgeActionButtonLabel
+            self.isCritical = isCritical
+        }
+    }
+    public struct Identifier: Equatable {
+        /// Unique device manager identifier from whence the alert came, and to which alert acknowledgements should be directed.
+        public let deviceManagerInstanceIdentifier: String
+        /// Per-alert unique identifier, for instance to group alert types.  This is the identifier that will be used to acknowledge the alert.
+        public let typeIdentifier: TypeIdentifier
```

Comment:
alertTypeIdentifier?

Reply:
I felt that it was kind of redundant, since this already was in a subtype of `UserAlert`.